nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -8,11 +8,161 @@
+        )
+
+    return_first = isinstance(test_value, (float, int))
+    test_value = np.atleast_1d(test_value)
+    null_array = np.array(null_array)
+
+    # For efficiency's sake, if there are more than 1000 values, pass only the
+    # unique values through percentileofscore(), and then reconstruct.
+    if len(test_value) > 1000:
+        reconstruct = True
+        test_value, uniq_idx = np.unique(test_value, return_inverse=True)
+    else:
+        reconstruct = False
+
+    def compute_p(t, null):
+        null = np.sort(null)
+        idx = np.searchsorted(null, t, side='left').astype(float)
+        return 1 - idx / len(null)
+
+    if tail == 'two':
+        if symmetric:
+            p = compute_p(np.abs(test_value), np.abs(null_array))
+        else:
+            p_l = compute_p(test_value, null_array)
+            p_r = compute_p(test_value * -1, null_array * -1)
+            p = 2 * np.minimum(p_l, p_r)
+    elif tail == 'lower':
+        p = compute_p(test_value * -1, null_array * -1)
+    else:
+        p = compute_p(test_value, null_array)
+
+    # ensure p_value in the following range:
+    # smallest_value <= p_value <= (1.0 - smallest_value)
+    smallest_value = np.maximum(np.finfo(float).eps, 1.0 / len(null_array))
+    result = np.maximum(smallest_value, np.minimum(p, 1.0 - smallest_value))
+
+    if reconstruct:
+        result = result[uniq_idx]
+
+    return result[0] if return_first else result
+
+
+def _calculate_cluster_measures(arr4d, threshold, conn, two_sided_test=False):
+    """Calculate maximum cluster mass and size for an array.
+
+    This method assesses both positive and negative clusters.
+
+    Parameters
+    ----------
+    arr4d : :obj:`numpy.ndarray`
+        Unthresholded 4D array of 3D t-statistic maps.
+    threshold : :obj:`float`
+        Uncorrected t-statistic threshold for defining clusters.
+    conn : :obj:`numpy.ndarray` of shape (3, 3, 3)
+        Connectivity matrix for defining clusters.
+
+    Returns
+    -------
+    max_size, max_mass : :obj:`float`
+        Maximum cluster size and mass from the matrix.
+    """
+    n_regressors = arr4d.shape[3]
+
+    max_sizes = np.zeros(n_regressors, int)
+    max_masses = np.zeros(n_regressors, float)
+
+    for i_regressor in range(n_regressors):
+        arr3d = arr4d[..., i_regressor]
+
+        if two_sided_test:
+            arr3d[np.abs(arr3d) <= threshold] = 0
+        else:
+            arr3d[arr3d <= threshold] = 0
+
+        labeled_arr3d, _ = ndimage.measurements.label(arr3d > 0, conn)
+
+        if two_sided_test:
+            # Label positive and negative clusters separately
+            n_positive_clusters = np.max(labeled_arr3d)
+            temp_labeled_arr3d, _ = ndimage.measurements.label(arr3d < 0, conn)
+            temp_labeled_arr3d[temp_labeled_arr3d > 0] += n_positive_clusters
+            labeled_arr3d = labeled_arr3d + temp_labeled_arr3d
+            del temp_labeled_arr3d
```

Comment:
This distinguishes negative and positive values in cluster assignment. However, I have read in a few places that a two sided test should retain the positive and negative maximum values in _separate_ null distributions. For example, in [Albajes-Eizagirre et al. (2019)](https://pubmed.ncbi.nlm.nih.gov/30389629/):

>SDM-PSI can currently save four different maximum statistics from the image of z-values: the largest z-value (i.e., voxel-based statistics), the maximum cluster size or mass after thresholding with a used-defined z- value (i.e., cluster-size or mass statistics) (Bullmore et al., 1999), and the maximum TFCE (Smith and Nichols, 2009). We have implemented TFCE in our software so that users who do not have FSL installed will still be able to use TFCE in SDM-PSI.
>This procedure theoretically only tests hypothesis in one direction (e.g., patients > controls), but not the hypothesis in the other direction (e.g., patients < controls), and thus, the procedure should be conducted twice, one for each direction. However, given that the hypotheses are complementary and a permutation test is computationally consuming, SDM-PSI saves two numbers from each permutation: one for the positive hypothesis (e.g., the highest z-value) and one for the negative hypothesis (e.g., the lowest z-value), in two separate null distributions.

Should the vFWE, csFWE, and cmFWE null distributions be split by sign?

Reply:
@bthirion do you think tracking positive and negative distributions separately is necessary, or is it related to https://github.com/nilearn/nilearn/pull/3181#discussion_r840718578 in that we should assume the null distribution is zero-centered and symmetric, and just the maximum absolute value?

Reply:
No, I would consider only the absolute value. I'm not aware of any case with two-sided tests where asymmetry would matter.

Reply:
Note that in he above paper if the null distribution is obtained by permutation, it has to be symmetric ...

Reply:
At least in ALE meta-analyses, the test statistic scales with sample size, so comparing two groups with different numbers of studies will naturally lead to a null distribution that is not zero-centered. Though, I suppose since z-statistics were used in that case, it would end up being symmetric.